src/Processors/Formats/Impl/NpyRowInputFormat.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+#include <Core/Block.h>
+#include <Processors/Formats/IRowInputFormat.h>
+#include <Processors/Formats/ISchemaReader.h>
+#include <Formats/FormatSettings.h>
+#include <Common/HashTable/HashMap.h>
+#include "Columns/IColumn.h"
+#include <Core/Field.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/Types.h>
+
+using NpySizeT = uint32_t;
+static const uint8_t NPY_DOCUMENT_END = 0x00;
+
+namespace DB
+{
+
+class ReadBuffer;
+
+class NpyRowInputFormat final : public IRowInputFormat
+{
+public:
+    NpyRowInputFormat(ReadBuffer & in_, Block header_, Params params_);
+
+    String getName() const override { return "NpyRowInputFormat"; }
+
+    void resetParser() override;
+
+private:
+    bool readRow(MutableColumns & columns, RowReadExtension &) override;
+    void readData(MutableColumns & columns);
+
+    void readRows(MutableColumns & columns);
+
+    void readValueAndinsertIntoColumn(MutableColumnPtr column);
+
+    std::unordered_map<String, String> header;
+    std::vector<int> shape;
+    DataTypePtr nestedType;
+    int endian;
+    int sizeForStrings = 0;
```

Comment:
Numpy data type is something statefull, it contains endiannes, type name and size of the data (and for more complex types that we can support in future, there can be more states). So, maybe we can create a set of classes for numpy data type that will store state according to the type.
For example:
```

enum class NumpyDataTypeIndex
{
    Int,
    Float,
    String,
    ...
};

class NumpyDataType
{
public:
    enum Endianness
    {
        LITTLE,
        BIG,
        NONE,
    };

    NumpyDataType(Endianness endianess_) : endianess(endianess_) {}
    virtual ~NumpyDataType() = default;

    Endianness getEndianness() const { return endianess; }

    virtual NumpyDataTypeIndex getTypeIndex() const = 0;

private:
    Endianness endianess;
};

class NumpyIntDataType : public NumpyDataType
{
public:
    NumpyIntDataType(Endianness endianess, size_t size_, bool is_signed_) : NumpyDataType(endianess), size(size_), is_signed(is_signed_) {}
    
    NumpyDataTypeIndex getTypeIndex() const override { return NumpyDataTypeIndex::Int; }
    size_t getSize() const { return size; }
    bool isSigned() const { return is_signed; }

private:
    size_t size;
    bool is_signed;
};

class NumpyFloatDataType : public NumpyDataType
{
public:
    NumpyFloatDataType(Endianness endianess, size_t size_) : NumpyDataType(endianess), size(size_) {}
    
    NumpyDataTypeIndex getTypeIndex() const override { return NumpyDataTypeIndex::Float; }
    size_t getSize() const { return size; }
    
private:
    size_t size;
};

class NumpyStringDataType : public NumpyDataType
{
public:
    NumpyStringDataType(Endianness endianess, size_t size_) : NumpyDataType(endianess), size(size_) {}

    size_t getSize() const { return size; }
private:
    size_t size;
};

...
```

So, we can create a shared_ptr of such type during dtype parsing. And during values reading we can use NumpyDataTypeIndex to check what type we have in data, cast our shared_ptr to the corresponding data type and use it's state (like endiness, size, is value signed or not, etc).
It can be moved to a separated file like `srd/Formats/NumpyDataTypes.h`.